File: py/genetic_algorithm_optimizer_test.py
import py.target_function as tf
import py.genetic_algorithm_optimizer as gao
import numpy as np


class TestTargetFunction(tf.TargetFunction):

    # f(x) = a * x + b
    def fun(self, params, args):
        x = args[0]
        a = params[0]
        b = params[1]
        res = a * x + b
        return res


def test_optimizer():
    bounaries = np.array([[-5, 5], [-5, 5]])
    optimizer = gao.GeneticAlgorithmOptimizer(bounaries)
    params = np.array([0.1, 0.1])
    train_input = np.array([[-1], [0], [1]])
    truth_output = np.array([[-1], [1], [3]])
    target_function = TestTargetFunction()
    optimized_params = optimizer.optimize(target_function, params, train_input, truth_output)
    print(optimized_params)


if __name__ == "__main__":
    test_optimizer()

File: src/test/java/jm/app/optimize/GeneticAlgorithmOptimizerTest.java
package jm.app.optimize;

import jm.app.algebra.Matrix;
import jm.app.optimize.GeneticAlgorithmOptimizer;
import jm.app.optimize.Optimizer;
import jm.app.optimize.TargetFunction;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class GeneticAlgorithmOptimizerTest {

    private Optimizer optimizer;

    @BeforeEach
    public void initOptimizer() {
        Matrix boundaries = new Matrix(2, 2);
        boundaries.setValue(0, 0, -5);
        boundaries.setValue(0, 1, 5);
        boundaries.setValue(1, 0, -5);
        boundaries.setValue(1, 1, 5);
        optimizer = new GeneticAlgorithmOptimizer(boundaries);
    }

    @Test
    public void testOptimize() {
        Matrix params = new Matrix(1, 2);
        params.setValue(0, 0, 0.01);
        params.setValue(0, 1, 0.01);

        Matrix trainInput = new Matrix(3, 1);
        trainInput.setValue(0, 0, -1);
        trainInput.setValue(1, 0, 0);
        trainInput.setValue(2, 0, 1);

        Matrix truthOutput = new Matrix(3, 1);
        truthOutput.setValue(0, 0, -1);
        truthOutput.setValue(1, 0, 1);
        truthOutput.setValue(2, 0, 3);

        Matrix optimizedParams = optimizer.optimize(new TargetFunction() {
            @Override
            public BigDecimal fun(Matrix params, Matrix args) {
                BigDecimal x = args.getValue(0, 0);
                BigDecimal a = params.getValue(0, 0);
                BigDecimal b = params.getValue(0, 1);
                BigDecimal res = a.multiply(x).add(b);
                return res;
            }
        }, params, trainInput, truthOutput);
        System.out.println(optimizedParams);
        final BigDecimal error = new BigDecimal(0.01);
        Assertions.assertTrue(optimizedParams.getValue(0, 0).subtract(new BigDecimal(2.0)).abs().
                subtract(error).compareTo(new BigDecimal(0.0)) < 0);
        Assertions.assertTrue(optimizedParams.getValue(0, 1).subtract(new BigDecimal(1.0)).abs().
                subtract(error).compareTo(new BigDecimal(0.0)) < 0);
    }
}
